languages/thingtalk/ast_manip.ts

```diff
@@ -1467,54 +1401,50 @@ function addGetPredicateJoin(table : Ast.Expression,
         return addFilterInternal(table, newAtom, {});
     }
 
-    let newAtom = new Ast.BooleanExpression.Atom(null, 'id', (lhsArg.type.isArray ? 'in_array' : '=='), new Ast.Value.VarRef(lhsArg.name));
-    if (negate)
-        newAtom = new Ast.BooleanExpression.Not(null, newAtom);
-    const get_predicate = new Ast.BooleanExpression.External(null,
-        get_predicate_table.expression.invocation.selector,
-        get_predicate_table.expression.invocation.channel,
-        get_predicate_table.expression.invocation.in_params,
-        new Ast.BooleanExpression.And(null, [get_predicate_table.filter, newAtom]),
-        get_predicate_table.expression.invocation.schema
+    // add id projection to subquery
+    if (subquery instanceof Ast.ProjectionExpression)
+        subquery.args = ['id'];
+    else
+        subquery = new Ast.ProjectionExpression(null, subquery, ['id'], [], [], subquery.schema);
```

Comment:
Wait is this the right form? `[id]` is a minimal projection, wouldn't it be optimized out if projected explicitly?

Reply:
Seem not the case. where is the code to check minimal projection? didn't find it in optimize.ts

Reply:
It might be a normalization that we apply only to synthesized data, and not to all programs. In which case it's fine if this the normalized form you want.